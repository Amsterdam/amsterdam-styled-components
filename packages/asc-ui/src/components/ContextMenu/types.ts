enum Position {
  bottom = 'bottom',
  top = 'top',
}

export default {
  Position,
}
