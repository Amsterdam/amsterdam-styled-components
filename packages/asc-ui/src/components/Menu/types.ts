import { Theme } from '@datapunt/asc-core'

export namespace MenuStyleProps {
  export enum Position {
    bottom = 'bottom',
    top = 'top',
  }

  export enum Align {
    left = 'left',
    right = 'right',
  }

  export type SharedStyleProps = {
    focused?: boolean
    height?: number
  }

  export type MenuButtonStyleProps = {} & SharedStyleProps

  export type MenuItemStyleProps = {
    role?: string
    id?: string
    label?: string
    open?: boolean
    divider?: boolean
    hoverColor?: Theme.TypeLevel
  } & SharedStyleProps

  export type MenuListStyleProps = {
    labelId?: string
    position?: Position
    right?: number
    top?: number
    focused?: boolean
    alignRight?: boolean
    css?: string
  }

  export type MenuWrapperStyleProps = {
    css?: string
  } & SharedStyleProps
}
