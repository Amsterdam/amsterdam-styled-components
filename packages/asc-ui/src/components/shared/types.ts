export interface LoadingInterface {
  loading?: boolean
}

export interface PageLayoutInterface {
  tall?: boolean
  fullWidth?: boolean
}
