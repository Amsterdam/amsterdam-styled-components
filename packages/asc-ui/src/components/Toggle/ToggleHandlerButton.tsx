import React from 'react'
import { Close, Menu } from '@datapunt/asc-assets'
import Icon from '../Icon'
import ButtonToggle from '../ButtonToggle/ButtonToggleStyle'
import { ToggleHandlerProps } from './Toggle'

const ToggleHandlerButton: React.FC<ToggleHandlerProps> = ({
  iconOpen,
  iconClose,
  open,
  onClick,
}) => {
  const IconOpen = iconOpen || <Menu />
  const IconClose = iconClose || <Close />
  return (
    <ButtonToggle open={!!open} onClick={onClick}>
      <Icon>{open ? IconClose : IconOpen}</Icon>
    </ButtonToggle>
  )
}

export default ToggleHandlerButton
