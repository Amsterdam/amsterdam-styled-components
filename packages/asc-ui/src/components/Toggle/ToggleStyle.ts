import styled from 'styled-components'
import { themeColor } from '../../utils/themeUtils'

export type Props = {
  id: string
  left: string
  right: string
  value: string
  className?: string
}

export const WrapperStyle = styled.span`
  input:checked + label {
    color: white;
    background-color: ${themeColor('secondary')};
    border: 2px solid ${themeColor('secondary')};
    padding: 12px 18px;
  }
`

type RadioProps = {
  role: string
  id: string
  type: string
}

export default styled.input.attrs({
  type: 'radio',
})<RadioProps>`
  opacity: 0;
  position: absolute;
`
// position: absolute Firefox error

type LabelProps = {
  htmlFor: string
}

export const LabelStyle = styled.label<LabelProps>`
  background-color: ${themeColor('tint', 'level3')};
  font-weight: bold;
  padding: 12px 20px;
`
