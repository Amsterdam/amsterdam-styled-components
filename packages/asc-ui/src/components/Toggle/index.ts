export {
  default as Toggle,
  ToggleHandlerProps,
  Props as ToggleProps,
} from './Toggle'
