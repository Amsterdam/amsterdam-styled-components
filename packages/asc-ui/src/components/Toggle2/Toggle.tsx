import React, {
  forwardRef,
  HTMLAttributes,
  useImperativeHandle,
  useRef,
} from 'react'
import SwitchStyle, {
  Props,
  LabelStyle,
  KnobStyle,
  WrapperStyle,
} from './ToggleStyle'

const Switch = forwardRef<
  HTMLInputElement,
  Props & HTMLAttributes<HTMLInputElement>
>(({ id, className, ...otherProps }, externalRef) => {
  const ref = useRef<HTMLInputElement>(null)

  useImperativeHandle(externalRef, () => ref.current as HTMLInputElement)

  return (
    <WrapperStyle className={className}>
      <LabelStyle htmlFor={id}>
        <SwitchStyle role="switch" {...{ ...otherProps, id, ref }} />
        <KnobStyle />
      </LabelStyle>
    </WrapperStyle>
  )
})

export default Switch
