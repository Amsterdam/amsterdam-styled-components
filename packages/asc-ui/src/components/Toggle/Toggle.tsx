import React from 'react'
import { Close, Menu } from '@datapunt/asc-assets'
import ToggleStyle, { Props as ToggleStyleProps } from './ToggleStyle'
import Icon from '../Icon'
import ButtonToggle from '../ButtonToggle/ButtonToggleStyle'
import ownerDocument from '../../utils/ownerDocument'
import usePassPropsToChildren from '../../utils/usePassPropsToChildren'
import useActionOnEscape from '../../utils/useActionOnEscape'

export type Props = {
  icon?: React.ReactElement
  open?: boolean
  render?: boolean
  onOpen?: Function
  as?: keyof JSX.IntrinsicElements | React.ComponentType<any>
} & React.HTMLProps<HTMLElement> &
  ToggleStyleProps

const Toggle: React.FC<Props> = ({
  children: childrenProps,
  onClick,
  open: openProp,
  onKeyDown,
  onOpen,
  css,
  render,
  icon,
  ...otherProps
}) => {
  const [open, setOpen] = React.useState(false)
  const { onKeyDown: onKeyDownHook } = useActionOnEscape(() => setOpen(false))
  const ref = React.useRef(null!)

  const handleOnBlur = () => {
    setTimeout(() => {
      const element = ref.current
      if (element) {
        const currentFocus = ownerDocument(element).activeElement
        // @ts-ignore
        if (!element.contains(currentFocus)) {
          setOpen(false)
        }
      }
    })
  }

  const handleOnKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    onKeyDownHook(e)
    if (onKeyDown) {
      onKeyDown(e)
    }
  }

  const handleOnClick = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    setOpen(!open)
    if (onClick) {
      onClick(e)
    }
  }

  const { children } = usePassPropsToChildren(childrenProps, {
    'aria-hidden': !open,
  })

  // Useful if parent needs to take over control the open state
  React.useEffect(() => {
    if (openProp) {
      setOpen(openProp)
    }
  }, [openProp])

  React.useEffect(() => {
    if (onOpen) {
      onOpen(open)
    }
  }, [open])

  const ClosedIcon = icon || <Menu />

  const conditionalRenderedChildren = open ? children : null

  return (
    <ToggleStyle
      // @ts-ignore
      ref={ref}
      css={css}
      onBlur={handleOnBlur}
      onKeyDown={handleOnKeyDown}
      {...otherProps}
    >
      <ButtonToggle open={open} onClick={handleOnClick}>
        <Icon>{open ? <Close /> : ClosedIcon}</Icon>
      </ButtonToggle>
      {render ? children : conditionalRenderedChildren}
    </ToggleStyle>
  )
}

Toggle.defaultProps = {
  render: true,
}

export default Toggle
