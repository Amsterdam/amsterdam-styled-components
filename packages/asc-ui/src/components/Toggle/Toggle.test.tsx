import { render, fireEvent } from '@testing-library/react'
import Toggle from './Toggle'
import { KeyboardKeys } from '../../types'

describe('Toggle', () => {
  let container: any
  let queryByTestId: any
  let button: HTMLButtonElement
  const onClickMock = jest.fn()
  beforeEach(() => {
    ;({ container, queryByTestId } = render(
      <Toggle onClick={onClickMock} render={false}>
        <div data-testid="child">Foo</div>
      </Toggle>,
    ))
    button = container.querySelector('button')
  })

  it('should render the toggle button', () => {
    expect(container.querySelector('button')).toBeDefined()
  })

  it('should display the children when the button is clicked', () => {
    expect(queryByTestId('child')).toBeNull()
    fireEvent.click(button)
    expect(queryByTestId('child')).toBeDefined()
    expect(onClickMock).toHaveBeenCalled()
  })

  it('should not render the children when user presses the escape key after opening it', () => {
    const escape = {
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      preventDefault: () => {},
      key: KeyboardKeys.Escape,
    }

    fireEvent.click(button)
    expect(queryByTestId('child')).toBeDefined()
    fireEvent.keyDown(button, escape)
    expect(queryByTestId('child')).toBeNull()
  })

  it('should call the passed parent function when the button is clicked', () => {
    const onOpenMock = jest.fn()
    ;({ container } = render(
      <Toggle onClick={jest.fn()} onOpen={onOpenMock} render={false}>
        <div id="child">Foo</div>
      </Toggle>,
    ))

    fireEvent.click(container.querySelector('button'))
    expect(onOpenMock).toHaveBeenCalledWith(true)
  })

  it('should display the children when the parent passes an open prop', () => {
    ;({ container } = render(
      <Toggle onClick={jest.fn()} open render={false}>
        <div id="child">Foo</div>
      </Toggle>,
    ))

    expect(queryByTestId('child')).toBeDefined()
  })
})
