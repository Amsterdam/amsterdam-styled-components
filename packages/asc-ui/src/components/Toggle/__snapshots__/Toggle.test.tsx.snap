// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 20px;
  height: 20px;
  box-sizing: content-box;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c2 > svg {
  width: inherit;
  height: inherit;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  cursor: pointer;
  font-size: 16px;
  font-weight: 500;
  line-height: 20px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 50px;
  width: 50px;
  -webkit-transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
}

.c1:focus {
  z-index: 10;
  outline-style: solid;
  outline-offset: 0px;
  outline-width: 3px;
}

.c1::after {
  content: '';
  display: block;
  min-height: inherit;
  font-size: 0;
}

.c1:disabled {
  cursor: default;
  outline: none;
  border: none;
}

<div
  class="c0"
>
  <button
    class="c1"
    size="50"
  >
    <span
      class="c2"
    >
      <svg
        focusable="false"
        viewBox="0 0 32 32"
      >
        <path
          d="M0 3.238h32V7.81H0V3.238zm0 10.476h32v4.572H0v-4.572zM0 24.19h32v4.572H0V24.19z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </button>
  <div
    aria-hidden="true"
    id="child"
  >
    Foo
  </div>
</div>
`;
